### PR TITLE
Add NonRetryableInternalService error and Error.IsUnexpected

### PR DESCRIPTION
**Pull request by tdinucci**
_This is an alternative to https://github.com/monzo/terrors/pull/48 and based on [this discussion](https://monzo.slack.com/archives/CNDDNNFTK/p1677695759034219)._

This PR adds a couple of things:

## NonRetryableInternalService

`internal_service` terrors are considered to be retryable by default.  It is possible to update a terror, marking it as not retryable however this is quite clunky.  

This PR adds a new factory function called `NonRetryableInternalService` which returns an `internal_service` terror which is not considered to be retryable.  It also adds an `Error.SetIsRetryable(bool)` method to make it easier to manipulate retryability on general terrors.

## Error.IsUnexpected

We have centralised logic which acts on errors, e.g. logging them.  In a small number of cases it is not clear how serious an error is, e.g. is it due to some benign, expected event or is it due to some defensive check which is never expected to fail (signifying a bug).

With this PR it's now possible to specify whether an error is unexpected.  Code which interprets unexpected errors will then be able to deal with them more appropriately.

### CODE REVIEW

File: errors.go [hunk 1]
```diff
@@ -160,6 +163,32 @@ func (p *Error) Retryable() bool {
...
+
+func (p *Error) SetIsRetryable(value bool) {
+	if value {
+		p.IsRetryable = &retryable
+	} else {
+		p.IsRetryable = &notRetryable
+	}
+}
+
+func (p *Error) SetIsUnexpected(value bool) {
```
Comment (danielchatfield) [inline errors.go:184]:
```suggestion
// SetIsUnexpected can be used to explicitly mark an error as unexpected or not. In practice the vast majority of 
// code should not need to use this. An example use case might be when returning a validation error that must
// mean there is a coding mistake somewhere (e.g. default statement in a switch that is never expected to be 
// taken). By marking the error as unexpected there is a greater chance that an alert will be sent.
func (p *Error) SetIsUnexpected(value bool) {
```

**Review by danielchatfield — APPROVED** [review]

**Comment by rdingwall** [comment]
`NonRetryableInternalServiceError` seems like an sensible improvement 👍 

I don't understand what `IsUnexpected` is for though. Can you provide some examples of when you would mark and error as `IsUnexpected` and how that would impact how those errors are handled?

**Review by rdingwall — COMMENTED** [review]
`NonRetryableInternalServiceError` seems like an sensible improvement 👍

I don't understand what `IsUnexpected` is for though. Can you provide some examples of when you would mark and error as unexpected and how that would impact how those errors are handled?

**Comment by tdinucci** [comment]
@rdingwall there is some discussion around `IsUnexpected` in the Slack thread that's linked above.

The underlying idea is that at the moment, in a small number of cases there can be ambiguity around the severity of an error.  For example, a `precondition_failed` error may in some cases be benign and expected, e.g. someone doesn't have adequate funds to complete a transfer.  In other cases a `precondition_failed` error may be returned because of some unexpected and dangerous event, e.g. a defensive validation check which is always expected to pass fails, indicating a bug somewhere.

This PR just adds the ability to mark errors as unexpected.  Higher level code, e.g. in the bedrock `ErrorLoggingFilter`, will be able to interpret this new field and react accordingly, e.g. by ensuring that the logged message severity is not downgraded.

**Comment by rdingwall** [comment]
> @rdingwall there is some discussion around `IsUnexpected` in the Slack thread that's linked above.
> 
> The underlying idea is that at the moment, in a small number of cases there can be ambiguity around the severity of an error. For example, a `precondition_failed` error may in some cases be benign and expected, e.g. someone doesn't have adequate funds to complete a transfer. In other cases a `precondition_failed` error may be returned because of some unexpected and dangerous event, e.g. a defensive validation check which is always expected to pass fails, indicating a bug somewhere.
> 
> This PR just adds the ability to mark errors as unexpected. Higher level code, e.g. in the bedrock `ErrorLoggingFilter`, will be able to interpret this new field and react accordingly, e.g. by ensuring that the logged message severity is not downgraded.

Understood. So it's for suppressing errors which are expected as part of normal control flow. 👍 

The main risk I'd be worried about is accidentally suppressing serious errors. At the moment, the `IsUnexpected` flag implies that all existing errors are expected by default, and thus are prone to being suppressed. Have you considered inverting the flag to `IsExpected`, so services opt-in for suppressing spurious errors instead?

Secondly, how will you handle errors across service boundaries?  E.g. in the code below, an `account.not_found` error is expected during signup, so shouldn't be logged. But inside `service.account` we don't know that the account is not expected to exist yet, so we might want to log the error.


https://github.com/monzo/wearedev/blob/d9bff1f542edca543c1845de3b4458c6a2203921/service.personal-account-signup/handler/create_application.go#L179-L180

if all errors are considered expected by default.


Out of curiosity why wouldn't we invert it, so all errors are unexpected


I would propose inverting the flag to opt-in for muting i.e. `IsExpected`

Got it. It feels like this would be safer to apply the inverse, i.e. all errors are unexpected and should be logged by default, but for particular cases where error are being used for normal control flow (and are thus expected),

 How would you solve

**Comment by tdinucci** [comment]
> Understood. So it's for suppressing errors which are expected as part of normal control flow. 👍

At the moment we already suppress errors with the `ErrorLogginFilter` and within this context I'd view this as it giving us a way to un-suppress errors.  If an error code is currently suppressed by the filter but the error is marked as `IsUnexpected` then we'd go ahead and log this at a high severity.

> Secondly, how will you handle errors across service boundaries?

This particular PR won't help with this.  

There is another PR that has potential though (currently in review).  If you have questions about this specific requirement it'd be good to discuss on https://github.com/monzo/wearedev/pull/131513

**Pull request merged**